### PR TITLE
RDFPFN792026: recognize external effect synchronization

### PR DESCRIPTION
**Pull request by aidenybai**
## Why

React Bench source reconstruction confirmed five `no-event-handler` false positives on immutable main `86add142688fa951a456d86c006f3f8c7c36c070`:

- four error-state effects defer focus until the conditional dismiss control exists after commit
- one effect synchronizes handler-written message state into a `Map` held by a React ref

These effects synchronize React state with external DOM or mutable-ref state; they are not state-mediated event handlers.

## What

- recognize scheduler callbacks that exclusively focus a proven React ref
- recognize `Map` and `Set` mutations through a proven React ref with a matching collection initializer
- retain diagnostics for deferred prop callbacks, mixed focus/event work, opaque ref clients, and shadowed `useRef`
- add five React Bench regressions, four adversarial negative controls, five strict-fuzz corpus cases, and a patch changeset

## Exact replay

| Revision | Files | Parse errors | Diagnostics |
| --- | ---: | ---: | ---: |
| immutable main `86add1426` | 5 | 0 | 5 |
| this PR `b21dc5087` | 5 | 0 | 0 |

Removed diagnostics: 5. Added diagnostics: 0.

## Validation

- focused rule contracts and regressions: pass
- full plugin suite: 25,100 pass / 201 skip
- full repository tests: 15/15 tasks pass
- strict fuzz: `FUZZ_RULE=no-event-handler FUZZ_STRICT=1 FUZZ_ITERATIONS=500 nr fuzz`
- fuzz corpus: 192 pass / 782 skip
- full build: 9/9 tasks pass
- lint: pass (existing corpus warnings only)
- typecheck: 16/16 tasks pass
- format: pass
- JSON report smoke: pass
- post-change truffler duplicate-symbol search: no duplicate helper surface found

The bounded RDE run could not start because the eval environment lacks required `AXIOM_DATASET`; no scans were run. Daytona parity is intentionally deferred because the shared parity queue is paused and must not be interrupted.

---

> [!NOTE]
> **Medium Risk**
> Changes heuristic classification in a widely used lint rule; regressions and adversarial tests limit false negatives, but edge cases may still slip through.
> 
> **Overview**
> Reduces **`no-event-handler`** false positives by treating some effect consequents as **external synchronization** instead of “faked event handlers.”
> 
> **`consequentHasTransferableWork`** now skips scheduler callbacks (`setTimeout`, `requestAnimationFrame`, etc.) whose body only calls **`.focus()`** on a proven React ref, and skips **`Map`/`Set`** mutations on refs initialized with **`new Map()`** / **`new Set()`**. Prop-callback timers, mixed focus plus other work, opaque ref API calls, and shadowed **`useRef`** still report.
> 
> Adds five fuzz corpus regressions (React Bench shapes), matching unit tests, adversarial “still reports” cases, and a patch changeset for **`oxlint-plugin-react-doctor`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ff11123156a322d3131b4fa8140ff4d17c79c1d2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Final current-main parity

The immutable 2,000-project matrix replay is complete against current main `6b64dfaf9e973139d1df2d09f405c9d916e158a3` and prospective merge commit `14973352b99a0e772c538746fc70f77b17fe436d`. Both candidate and scoped shared-base artifacts contain 2,000/2,000 records with zero failed or skipped projects. The exact `react-doctor/no-event-handler` comparison is `+0/-0`, with 1,322 unchanged diagnostic identities.

- candidate NDJSON SHA-256: `2916e4568f71f3f56b2e5d6689b5be238bb0bd20ff705e8d518249d120bed14a`
- scoped base NDJSON SHA-256: `7e22605a84ef17853e0f2a748fc6fe2a11e3a1a0180ef190f25671f26140f4ee`
- parity JSON SHA-256: `23efbff491d524f36183967b9777f03d065e1ff06d75c71f1ffd0824faec0d61`
- descriptor SHA-256: `59e18f816a01b3c8367690c5173ef2e934d59c0015309de5a7efdb04b729f268`
- evaluator source SHA-256: `7dcb7c25d0608f37151cd4354d708f14980876a5463fcea95d4ad4d3f9f21c40`
- corpus manifest SHA-256: `ff8f5bfca41e411d4ad16b57d69c90c5a7b723ea6190f5ef94052af64af02b01`

The Group B evaluator was cleaned and its exact Daytona sandbox/snapshot set was verified absent.

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-event-handler.ts [hunk 1]
```diff
@@ -56,6 +172,16 @@ const consequentHasTransferableWork = (
     if (hasTransferableWork) return false;
     if (child !== consequent && isFunctionLike(child)) return false;
     if (!isNodeOfType(child, "CallExpression")) return;
+    if (
+      isDeferredRefFocusSynchronization(child, scopes) ||
+      isCollectionRefMutation(child, scopes)
+    ) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-event-handler.ts:174]: Fixed in ff1112315 by requiring a deferred-focus callback body to contain exactly one focus expression. Assignments, updates, additional statements, and other mixed work now keep the diagnostic. Added a regression control for focus plus document.title assignment; the plugin suite, package typecheck, strict 500-iteration fuzz, and exact five-source React Bench replay all pass.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-event-handler.ts [hunk 2]
```diff
@@ -56,6 +166,16 @@ const consequentHasTransferableWork = (
...
+    if (findTriggeredSideEffectCalleeName(child)) {
+      hasTransferableWork = true;
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-event-handler.ts:178]: No code change is needed here. The callback body is intentionally pruned, but the later local helper invocation is still visited as a CallExpression. For an identifier callee that is neither a state setter nor a custom-hook result, consequentHasTransferableWork falls through to hasTransferableWork = true. I replayed the proposed exact shape (const send = () => post(...); send()) against ff1112315: 0 parse errors and 1 no-event-handler diagnostic.
